### PR TITLE
Use API call to save instead of direct call

### LINKED ISSUE #388: Migrate Direct call to PluginJsonStorage to API.save()

### PR DESCRIPTION
**Pull request by taooceros**
close #388 

TODO:
- [x] Implement a auto call to Save all plugin's json storage that has registered
- [ ] Implement a better (safer) store to replace Dictionary<Type, object>
- [x] Finish up some settings

### CODE REVIEW

File: Flow.Launcher/PublicAPIInstance.cs [hunk 1]
```diff
@@ -133,15 +133,15 @@ public void OpenSettingDialog()
 
         public void LogException(string className, string message, Exception e, [CallerMemberName] string methodName = "") => Log.Exception(className, message, e, methodName);
 
-        private readonly Dictionary<Type, dynamic> PluginJsonStorages = new Dictionary<Type, dynamic>();
+        private readonly Dictionary<Type, object> PluginJsonStorages = new Dictionary<Type, object>();
```
Comment (taooceros) [inline Flow.Launcher/PublicAPIInstance.cs:136]: dynamic seems not working as intended, so I move back to direct cast

**Comment by jjw24** [comment]
Did you want me to review and merge, or you still working on the other two todos?

**Comment by taooceros** [comment]
> Did you want me to review and merge, or you still working on the other two todos?

still working on it. There are a few things that need to be cleaned up.
I am struggling with the second one, but it would be fine to not do it.

**Comment by taooceros** [comment]
I didn't find a better way to replace the Dictionary<Type, object> and reflection for registration. So just review the current one.

**Comment by jjw24** [comment]
Version bumps for plugins please :)

**Comment by taooceros** [comment]
> Version bumps for plugins please :)

Sure

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs [hunk 2]
```diff
@@ -113,11 +111,6 @@ public Control CreateSettingPanel()
             return new SettingsControl(_settings);
         }
 
-        public void Save()
-        {
-            _storage.Save();
-        }
-
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: why is this removed? settings still needs to be saved right

**Comment by jjw24** [comment]
Not related to this change but I think we should update the comment to give an example scenario where you will need to override the original class instance, i.e. use the override save method and when you just need to use save the original instance. Since you are in this space maybe you could just give a good explanation in the comment that would be awesome.
https://github.com/Flow-Launcher/Flow.Launcher/blob/b37f9f338aee9c2ecd5a50ae47cecd2fa4dae5cb/Flow.Launcher.Plugin/IPublicAPI.cs#L174-L188

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs (hunk 2 above)
Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: I add a auto save for all registered jsonstorage, so that we don't need to configure isavable for setting, unless they are going to save something else.

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: https://github.com/Flow-Launcher/Flow.Launcher/blob/b37f9f338aee9c2ecd5a50ae47cecd2fa4dae5cb/Flow.Launcher/PublicAPIInstance.cs#L139-L146

**Comment by taooceros** [comment]
I have updated the comment. In fact, I don't believe we may actually need the one to overwrite it, because the supposed process is to load and use the loaded file (which will be a new when none exist), and save the original one. Should we remove it?

Also, do you think we should modify the name specifically to setting because the file will be stored inside the setting folder.

By the way, would you please give me the access for OAuth from Jetbrains IDE integration? Sometimes I may use rider instead of Visual Studio or vscode. Thank you!

**Comment by jjw24** [comment]
> By the way, would you please give me the access for OAuth from Jetbrains IDE integration? Sometimes I may use rider instead of Visual Studio or vscode. Thank you!

Done

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs (hunk 2 above)
Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: 1. would this not be a doubling up then, with PluginManager.Save() call?

2. also you need to move this SaveAllJsonStorage() call to here right so when Flow's settings window closed you want to save all plugin settings:
https://github.com/Flow-Launcher/Flow.Launcher/blob/9d4af83fef727a38e22daf6e636351f42e3e1c08/Flow.Launcher/ViewModel/SettingWindowViewModel.cs#L85-L97

Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: 3. do we want to remove ISavable?
4. if it's auto saved, might want to document it somewhere right? for those that develop plugins

**Comment by jjw24** [comment]
> I have updated the comment. In fact, I don't believe we may actually need the one to overwrite it, because the supposed process is to load and use the loaded file (which will be a new when none exist), and save the original one. Should we remove it?

i am unsure when will a circumstance be needed to save a new instance to the original, so i would say yes

> Also, do you think we should modify the name specifically to setting because the file will be stored inside the setting folder.

yes

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs (hunk 2 above)
Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: > would this not be a doubling up then, with PluginManager.Save() call?

I guess that one will call Save from ISavable? So I remove the call to save in exist plugin.

> also you need to move this SaveAllJsonStorage() call to here right so when Flow's settings window closed you want to save all plugin settings:

Yeah you are right.

> if it's auto saved, might want to document it somewhere right? for those that develop plugins

Yeah, I forget to add that into the api description

**Comment by taooceros** [comment]
> > By the way, would you please give me the access for OAuth from Jetbrains IDE integration? Sometimes I may use rider instead of Visual Studio or vscode. Thank you!
> 
> Done

Thank you!

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs (hunk 2 above)
Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: > do we want to remove ISavable?

I believe not. There are other things that may need to save than setting (e.g. the program cache in program plugin).

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: image

I guess we don't need to add it to viewmodel save. The api save will be called once Flow is disposed.

Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: > > do we want to remove ISavable?
> 
> I believe not. There are other things that may need to save than setting (e.g. the program cache in program plugin).

yeah you are right, lets keep it as we do need to save non-settings, but lets document the Save() method in so it's clear, otherwise devs might use it thinking that's how you save settings.

Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: > image
> 
> I guess we don't need to add it to viewmodel save. The api save will be called once Flow is disposed.

We need to because it's better to save when user close the settings window, otherwise what of the program crashes, no settings will be saved.

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: Got you

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: > We need to because it's better to save when user close the settings window, otherwise what if the program crashes, no settings will be saved.

Done

File: Flow.Launcher.Plugin/IPublicAPI.cs [hunk 3]
```diff
@@ -34,6 +34,11 @@ public interface IPublicAPI
         /// </summary>
         void SaveAppAllSettings();
 
+        /// <summary>
+        /// Save Flow's plugins settings
+        /// </summary>
+        void SavePluginSettings();
```
Comment (jjw24) [inline Flow.Launcher.Plugin/IPublicAPI.cs:40]: what's the difference between this method and SaveSettingJsonStorage()?

File: Flow.Launcher.Plugin/IPublicAPI.cs [hunk 4]
```diff
@@ -165,33 +170,20 @@ public interface IPublicAPI
...
         /// <typeparam name="T">Type for deserialization</typeparam>
         /// <returns></returns>
-        T LoadJsonStorage<T>() where T : new();
+        T LoadSettingJsonStorage<T>() where T : new();
 
         /// <summary>
-        /// Save JsonStorage for current plugin. This is the method used to save settings to json in Flow.Launcher
+        /// Save JsonStorage for current plugin's setting. This is the method used to save settings to json in Flow.Launcher
         /// This method will save the original instance loaded with LoadJsonStorage.
+        /// This API call is for manually Save. Flow will automatically save all setting that has registered.
```
Comment (jjw24) [inline Flow.Launcher.Plugin/IPublicAPI.cs:183]: i think it will be better to also include how setting gets registered in the comment so plugin devs do not need to dig into code to find out.

File: Flow.Launcher.Plugin/IPublicAPI.cs (hunk 3 above)
Comment (taooceros) [reply Flow.Launcher.Plugin/IPublicAPI.cs:40]: SaveSettingJsonStorage is for a specific Type (or T as generic), but this is for all plugin.

File: Flow.Launcher.Plugin/IPublicAPI.cs [hunk 5]
```diff
@@ -26,11 +26,11 @@ public interface IPublicAPI
 
         /// <summary>
         /// Restart Flow Launcher
-        /// </summary>
+        /// </summar
```
Comment (taooceros) [inline Flow.Launcher.Plugin/IPublicAPI.cs:29]: is this typo?

Comment (jjw24) [reply Flow.Launcher.Plugin/IPublicAPI.cs]: yes. i fixed while i was updating the comments.

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs (hunk 2 above)
Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: can you add SaveAllJsonStorage() call to here so when Flow's settings window closed you want to save all plugin settings:
https://github.com/Flow-Launcher/Flow.Launcher/blob/9d4af83fef727a38e22daf6e636351f42e3e1c08/Flow.Launcher/ViewModel/SettingWindowViewModel.cs#L85-L97

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: I think this method is called in SaveAllJsonStorage, so it is impossible to directly call that method here, or else will become infinity recurse.
The call for 
```c#
PluginManager.Save(); 
_storage.Save(); 
```
do save all plugin setting.

Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: yeah i see. hmm how do save plugin settings when user closes settings window? PluginManager.Save() only saves plugin settings where the interface ISavable is applied and _storage.Save() saves Flow's own settings.

Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: instead of passing the api to the view, which is not even used, we can and should pass it to the viewmodel, this then allows you to call save plugin settings seperately from the view
https://github.com/Flow-Launcher/Flow.Launcher/blob/1e687a213039ba5f0c2fed42c7ef975a2d854141/Flow.Launcher/SettingWindow.xaml.cs#L28

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]: > PluginManager.Save() only saves plugin settings where the interface ISavable is applied and _storage.Save() saves Flow's own settings.

I have modified PluginManager.Save() to include saving plugins settings.

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs:120]:
```c#
public static void Save()
        {
            foreach (var plugin in AllPlugins)
            {
                var savable = plugin.Plugin as ISavable;
                savable?.Save();
            }
            API.SavePluginSettings();
        }
```

File: Plugins/Flow.Launcher.Plugin.Explorer/ViewModels/SettingsViewModel.cs [hunk 6]
```diff
@@ -9,27 +9,20 @@ namespace Flow.Launcher.Plugin.Explorer.ViewModels
...
-        public Task LoadStorage()
-        {
-            return Task.Run(() => Settings = storage.Load());
-        }
 
         public void Save()
         {
-            storage.Save();
+            Context.API.SaveSettingJsonStorage<Settings>();
         }
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Explorer/ViewModels/SettingsViewModel.cs:26]: one more question, why does this have to call the save directly?

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.Explorer/ViewModels/SettingsViewModel.cs:26]: I guess it is the design of Explorer plugin. I didn't change it. It contains a call for save every time a new item is edited.
It is not ISavable, so I believe it won't be called twice. However, do you think we should remove that?

Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.Explorer/ViewModels/SettingsViewModel.cs:26]: oh right, it's not ISavable, just calls infra's save directly. Nah don't worry about it for this PR, using the API is at least better. Let's get this merged in.

**Review by jjw24 — APPROVED** [review]
